src/nano_contracts/deserializer.ts

```diff
@@ -73,91 +93,129 @@ class Deserializer {
   /**
    * Deserialize string value
    *
-   * @param {value} Value to deserialize
+   * @param {Buffer} buf Value to deserialize
    *
    * @memberof Deserializer
    * @inner
    */
-  toString(value: Buffer): string {
-    return value.toString('utf8');
+  toString(buf: Buffer): DeserializeResult<string> {
+    // INFO: maxBytes is set to 3 becuase the max allowed length in bytes for a string is
+    // NC_ARGS_MAX_BYTES_LENGTH which is encoded as 3 bytes in leb128 unsigned.
+    // If we read a fourth byte we are definetely reading a higher number than allowed.
+    const {
+      value: lengthBN,
+      rest,
+      bytesRead: bytesReadForLength,
+    } = leb128Util.decodeUnsigned(buf, 3);
+    if (lengthBN > NC_ARGS_MAX_BYTES_LENGTH) {
+      throw new Error('String length in bytes is higher than max allowed');
+    }
+    // If lengthBN is lower than 64 KiB than its safe to convert to Number
+    const length = Number(lengthBN);
+    if (rest.length < length) {
+      throw new Error('Do not have enough bytes to read the expected length');
+    }
+    return {
+      value: rest.subarray(0, length).toString('utf8'),
+      bytesRead: length + bytesReadForLength,
+    };
   }
 
   /**
    * Deserialize bytes value
    *
-   * @param {value} Value to deserialize
+   * @param buf Value to deserialize
    *
    * @memberof Deserializer
    * @inner
    */
-  toBytes(value: Buffer): Buffer {
-    return value;
+  toBytes(buf: Buffer): DeserializeResult<Buffer> {
+    // INFO: maxBytes is set to 3 becuase the max allowed length in bytes for a string is
+    // NC_ARGS_MAX_BYTES_LENGTH which is encoded as 3 bytes in leb128 unsigned.
+    // If we read a fourth byte we are definetely reading a higher number than allowed.
+    const {
+      value: lengthBN,
+      rest,
+      bytesRead: bytesReadForLength,
+    } = leb128Util.decodeUnsigned(buf, 3);
+    if (lengthBN > BigInt(NC_ARGS_MAX_BYTES_LENGTH)) {
+      throw new Error('String length in bytes is higher than max allowed');
+    }
+    // If lengthBN is lower than 64 KiB than its safe to convert to Number
+    const length = Number(lengthBN);
+    if (rest.length < length) {
+      throw new Error('Do not have enough bytes to read the expected length');
+    }
+    return {
+      value: rest.subarray(0, length),
+      bytesRead: length + bytesReadForLength,
+    };
```

Comment:
This method is the same as `toString`, except for the `toString('utf-8')`. Should we use a single util method for it?

Reply:
Done